site/content/blog/2020-02-26-layered-store-and-struct-embedding.md

```diff
@@ -0,0 +1,335 @@
+---
+title: "Layered Store and struct embedding"
+slug: layered-store-and-struct-embedding
+date: 2020-02-26
+categories:
+    - "go"
+author: Jesús Espino
+github: jespino
+community: jesus.espino
+---
+
+One of the most important parts of the mattermost source code is the one
+responsible to accessing the mattermost database, the store. Every
+single database access is handled by the store, so we needed to find a way to
+extend it in different ways introducing as less complexity as possible,
+and that is the reason behind the current layered approach using struct embedding.
+
+Our store is the responsible for storing and retrieving data, and sometimes we
+need to add functionality that is not strictly related to the database queries,
+for example, cache data or add instrumentation. That are transversal tasks
+and not necesarilly need to be in the same block of code.
+
+Our approach is based in idea of a core Store, the one accessing to the source
+of truth (the database), and a set of layers on top of that adding extra
+behavior. All of them, the core Store and all the layers must implement
+the same interface, in this case the Store interface. This way, from the
+outside, there is not difference between our core Store without layers, or our
+core Store with 1000 layers on top of it.
+
+![layers](/blog/2020-02-26-layered-store-and-struct-embedding/layers.png)
+
+Each layer is going to embed another layer until the last layer embeds the core
+store. Each layer is going to override some methods (or all, depending on the
+layer). Each layer is responsible to decide what is handled entirely by the
+layer (for example a cache hit) or is delegated to the underneath layer (for
+example a cache miss).
+
+In our case, the core store is the SqlStore which encapsulates all the SQL
+queries execution. And the layers that we put on top of that are:
+
+  - The Cache Layer: responsible for maintaining a cache of the store calls.
+  - The Search Layer: responsible for speeding up the search methods using
+    search engines.
+  - The Timer Layer: responsible for sending the duration of each request to an
+    histogram in Prometheus.
+  - The OpenTracing Layer: responsible for of add the open tracing information
+    related to the store to the open tracing context.
+
+This 4 layers are adding functionality to our store but without touching a
+single line of code of the SqlStore implementation.
+
+How we do that?
+
+To solve this problem we used structure embedding what is the way go
```

Comment:
```suggestion
To solve this problem we used structure embedding which is what Go
```

Reply:
I believe the intent here is:
To solve this problem we used structure embedding which is the best practice in situations like this

Reply:
I wouldn't say is "the best practice", is not necessarily right or wrong, is the tool provided by go for extend functionality the behavior of a structure based on another.

Reply:
How about: 
"To solve this problem, we used structure embedding which is a tool provided by Go to extend the behavior of a struct based on another struct."